src/Middleware/HttpLogging/src/BufferingStream.cs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class BufferingStream : Stream, IBufferWriter<byte>
+    {
+        private const int MinimumBufferSize = 4096; // 4K
+        protected int _bytesBuffered;
+        private BufferSegment? _head;
+        private BufferSegment? _tail;
+        protected Memory<byte> _tailMemory; // remainder of tail memory
+        protected int _tailBytesBuffered;
+        protected ILogger _logger;
+        protected Stream _innerStream;
+
+        public BufferingStream(Stream innerStream, ILogger logger)
+        {
+            _logger = logger;
+            _innerStream = innerStream;
+        }
+
+        public override bool CanSeek => _innerStream.CanSeek;
+
+        public override bool CanRead => _innerStream.CanRead;
+
+        public override bool CanWrite => _innerStream.CanWrite;
+
+        public override long Length => _innerStream.Length;
+
+        public override long Position
+        {
+            get => _innerStream.Position;
+            set => _innerStream.Position = value;
+        }
+
+        public override int WriteTimeout
+        {
+            get => _innerStream.WriteTimeout;
+            set => _innerStream.WriteTimeout = value;
+        }
+
+        public string GetString(Encoding? encoding)
+        {
+            try
+            {
+                if (_head == null || _tail == null)
+                {
+                    // nothing written
+                    return "";
+                }
+
+                if (encoding == null)
+                {
+                    _logger.UnrecognizedMediaType();
+                    return "";
+                }
+
+                // Only place where we are actually using the buffered data.
+                // update tail here.
+                _tail.End = _tailBytesBuffered;
+
+                var ros = new ReadOnlySequence<byte>(_head, 0, _tail, _tailBytesBuffered);
+
+                var bufferWriter = new ArrayBufferWriter<char>();
```

Comment:
What's this for? It's also not pooled.

Reply:
For calling EncodingExtensions.Convert. It needs an IBufferWriter for the characters. 

I filed https://github.com/dotnet/aspnetcore/issues/31961 as follow up.